serviceregistration/kubernetes/retry_handler.go

```diff
@@ -22,58 +25,86 @@ type retryHandler struct {
 	// To synchronize setInitialState and patchesToRetry.
 	lock sync.Mutex
 
-	// setInitialState will be nil if this has been done successfully.
-	// It must be done before any patches are retried.
-	setInitialState func() error
+	// initialStateSet determines whether an initial state has been set
+	// successfully or whether a state already exists.
+	initialStateSet bool
+
+	// State stores an initial state to be set
+	initialState sr.State
 
 	// The map holds the path to the label being updated. It will only either
 	// not hold a particular label, or hold _the last_ state we were aware of.
 	// These should only be updated after initial state has been set.
 	patchesToRetry map[string]*client.Patch
-}
 
-func (r *retryHandler) SetInitialState(setInitialState func() error) {
-	r.lock.Lock()
-	defer r.lock.Unlock()
-	if err := setInitialState(); err != nil {
-		if r.logger.IsWarn() {
-			r.logger.Warn(fmt.Sprintf("unable to set initial state due to %s, will retry", err.Error()))
-		}
-		r.setInitialState = setInitialState
-	}
+	// client is the Client to use when making API calls against kubernetes
+	client *client.Client
 }
 
 // Run must be called for retries to be started.
-func (r *retryHandler) Run(shutdownCh <-chan struct{}, wait *sync.WaitGroup, c *client.Client) {
+func (r *retryHandler) Run(shutdownCh <-chan struct{}, wait *sync.WaitGroup) {
+	wait.Add(1)
+
+	r.setInitialState(shutdownCh)
+
 	// Run this in a go func so this call doesn't block.
 	go func() {
 		// Make sure Vault will give us time to finish up here.
-		wait.Add(1)
```

Comment:
Question, was there a reason this was moved up to the top?

Reply:
`wait.Add(1)` should always be called outside (and right before) the goroutine that you're waiting on (e.g. https://golang.org/pkg/sync/#example_WaitGroup). It should actually be somewhere around L49 -- I'll get that updated.

Reply:
Ah thanks! I was worried that someday, someone might add code error returning before the goroutine could ever be used to mark it done.